tools/rbd/action/Bench.cc: New options for rbd bench

bill-scales opened the pull request:
* Extend --io-type to support write_zeros and discard as well as read, write and readwrite
* Add --io-start to allow I/O to be started at a non-zero offset in the image

Splitting off rbd bench tool update into a separate PR from other change in https://github.com/ceph/ceph/pull/57738

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

idryomov commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -279,10 +298,10 @@ int do_bench(librbd::Image& image, io_type_t io_type,
       break;
     case IO_PATTERN_SEQ:
-      start_pos = seq_chunk_length * i;
+      start_pos = (io_start + seq_chunk_length * i) % size;
```
I don't think this handles the case when `--io-start` is set such that there is less than `--io-size` bytes available until the end of the image correctly -- `rbd_bencher_completion()` insists on being able to perform a full read at all times:
```
  } else if (b->io_type == IO_TYPE_READ && (unsigned int)ret != b->io_size) {
    std::cout << "read error: " << cpp_strerror(ret) << std::endl;
    exit(ret < 0 ? -ret : ret);
```

bill-scales replied:
Agree. There are few existing issues with IO_PATTERN_SEQ before adding io_start:

* If image size is small such that io_threads * io_size is > image size then seq_chunk_length becomes 0 and all the threads repeated issue I/O at pos 0. I think this is a bug, I'd expected that the threads would access the whole image.

* If io_bytes is greater than image size then after io_bytes of I/O each thread repeatedly issuing I/O to the same position at the start of their range. I think this is a bug, I'd expect that the positions would wrap and continue incrementing.

* If image size is not a multiple of seq_chunk_length then there is some special code to issue I/Os at the end of the image once the threads have each issued seq_chunk_length bytes of I/O. This means that I/Os are initially spaced seq_chunk_length apart, but there is then a small amount of sequential I/O at the end of the image. More a feature than a bug, I'd expected that the spacing between threads would be adjusted to deal with the remainder.

To make this work with io_start I'll make these changes:

* Add an input parameter validation constraint that io_start + io_size must be less than image size
* If io_threads * io_size > image size switch to IO_PATTERN_FULL_SEQ so that I/O occurs sequentially across the whole image
* If image size is not a multiple of seq_chunk_length then space some threads seq_chunk_length+io_size apart and the rest seq_chunk_length apart to deal with the remainder
* Fix the code so that after io_bytes has been accessed the I/O positions wrap and continue incrementing

With these changes it is then easy to adjust the start point for I/O

idryomov commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -279,10 +298,10 @@ int do_bench(librbd::Image& image, io_type_t io_type,
       break;
     case IO_PATTERN_FULL_SEQ:
-      start_pos = i * io_size;
+      start_pos = io_start + i * io_size;
```
... and here no attempt to wrap based on image size is made at all?

bill-scales replied:
Agree. Code was already missing a check that i * io_size was within the image size so I'll fix that as well.

bill-scales commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -162,7 +167,13 @@ struct rbd_bencher {
 					 rbd_bencher_completion);
-      image->aio_write2(off, len, write_bl, c, op_flags);
+      if (io_type == IO_TYPE_WRITE_ZEROES) {
+        image->aio_write_zeroes(off, len, c, RBD_WRITE_ZEROES_FLAG_THICK_PROVISION, op_flags);
```
I agree that the write_zeros is not very useful, especially with the other PR which lets you choose what pattern of data is being written (for testing Erasure Coding it can be quite useful to be able to write 0's to parts of stripe). I'll remove it from this PR and just add support for discard. I'll leave providing a more programmatic way of issuing I/Os and exercising the flags better to a future PR.

bill-scales commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -267,27 +305,23 @@ int do_bench(librbd::Image& image, io_type_t io_type,
-      start_pos = (rand() % (size / io_size)) * io_size;
-      break;
-    case IO_PATTERN_SEQ:
-      start_pos = seq_chunk_length * i;
```
Note strange behavior in old code when io_size * io_threads > size, seq_chunk_length becomes 0 and you do not get sequential I/O. This corner case is fixed below

bill-scales commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -267,27 +305,23 @@ int do_bench(librbd::Image& image, io_type_t io_type,
-      start_pos = seq_chunk_length * i;
-      break;
-    case IO_PATTERN_FULL_SEQ:
-      start_pos = i * io_size;
```
Note bug in old code when (io_size * io_threads) + io_size > size this will generate out of bounds I/Os

bill-scales commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -336,44 +370,40 @@ int do_bench(librbd::Image& image, io_type_t io_type,
-          thread_offset[i] += io_size;
-        } else {
-          // thread_offset is adjusted to the chunks unassigned to threads.
-          thread_offset[i] = off + (i * io_size);
```
Note strange behavior in old code where remainder (size % seq_chunk_length) is non-zero, this part at the end of image is accessed sequentially after rest of image. New code spaces threads at uneven intervals so whole image is accessed by io_thread sequential streams

bill-scales commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -336,44 +370,40 @@ int do_bench(librbd::Image& image, io_type_t io_type,
+          pos += io_size;
         }
-        if (thread_offset[i] + io_size > size) {
-          thread_offset[i] = seq_chunk_length * i;
```
Note bug in old code - if io_bytes is greater than image size then after sequentially accessing whole image once the threads get stuck repeatedly accessing this offset because off > size

idryomov commented on src/tools/rbd/action/Bench.cc:
```diff
@@ -336,44 +370,40 @@ int do_bench(librbd::Image& image, io_type_t io_type,
-          thread_offset[i] = off + (i * io_size);
+	// Advance by chunk length to get pos for next thread
+        pos += seq_chunk_length;
+        if (i < rem_chunks) {
```
Same question on assumptions for `i`.

idryomov commented:
@bill-scales There are a couple of outstanding questions and a (trivial) rebase is needed following the merge of Connor's change.  Would you have time to respin this?

bill-scales commented:
@idryomov appoligies for the delay, new version ready for review that addresses your comments. I did a lot more testing of the spacing of multi-threaded I/O with image sizes that were not a multiple of the number of threads after you found the bug with i being incremented too early and am confident that this works properly now.
